### PR TITLE
DownloadAgent

### PR DESCRIPTION
**Pull request by denisdimitrov**

### CODE REVIEW

File: src/test/java/downloadAgent/DownloadAgentTests.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Test
+    public void happyPath() throws Exception {
+        context.checking(new Expectations() {{
+            oneOf(downloadBar).get(100);
+        }});
+        DownloadAgent downloadAgent = new DownloadAgent(downloadBar, 10);
+        File actual = new File("actual.txt");
+        File expected = new File("expected.txt");
+        downloadAgent.downloadFile("File:src/test/java/downloadAgent/actual.txt", "src/test/java/downloadAgent/expected.txt");
```
Comment (mgenov) [inline src/test/java/downloadAgent/DownloadAgentTests.java]: You need to use class-loader for loading of test resources. 

https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getResourceAsStream(java.lang.String)


Keep in mind that test will fail if I run tests from different root folder.

File: src/main/java/downloadAgent/DownloadBar.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package downloadAgent;
+
+/**
+ * Created by clouway on 11.10.16.
```
Comment (mgenov) [inline src/main/java/downloadAgent/DownloadBar.java]: Created by need to be removed from the template.

File: src/main/java/downloadAgent/DownloadBar.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface DownloadBar {
+    void get(int percentage);
```
Comment (mgenov) [inline src/main/java/downloadAgent/DownloadBar.java]: I suspect that this is a listener class which listens for progress updates. 


ProgressChangeListener sounds much better and properly displays the intention. 
```java
public interface ProgressChangeListener {
  void onProgressUpdated(int newProgress); 
}
```

File: src/test/java/downloadAgent/DownloadAgentTests.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+public class DownloadAgentTests {
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery();
+
+    private DownloadBar downloadBar = context.mock(DownloadBar.class);
+
+    @Test
+    public void happyPath() throws Exception {
+        context.checking(new Expectations() {{
+            oneOf(downloadBar).get(100);
```
Comment (mgenov) [inline src/test/java/downloadAgent/DownloadAgentTests.java]: Progress is updated from 0 to 100 and I'm not able to see how you are verifying it.

File: src/test/java/downloadAgent/DownloadAgentTests.java (hunk 1 above)
Comment (denisdimitrov) [inline src/test/java/downloadAgent/DownloadAgentTests.java]: I tried using "InputStream is = DownloadAgentTests.class.getResourceAsStream("src/test/java/downloadAgent/actual.txt");" but it keeps returning NullPointerException, what am I doing wrong?

**Comment by denisdimitrov** [comment]
Fixed.

File: src/test/java/downloadAgent/DownloadAgentTests.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            oneOf(downloadBar).onProgressUpdated(90);
+            oneOf(downloadBar).onProgressUpdated(100);
+        }});
+        DownloadAgent downloadAgent = new DownloadAgent(downloadBar, 1);
+        File actual = new File("actual.txt");
+        File expected = new File("expected.txt");
+        URL url = DownloadAgentTests.class.getResource("actual.txt");
+        InputStream in1 = DownloadAgentTests.class.getResourceAsStream("expected.txt");
+        downloadAgent.downloadFile(url.toString(), in1.toString());
+        assertEquals(expected.length(), actual.length());
```
Comment (EmilGeorgiev) [inline src/test/java/downloadAgent/DownloadAgentTests.java]: The Length of files can be equal, but this that not guaranty that they have same content. We should assert bytes.

File: src/main/java/downloadAgent/DownloadAgent.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            int progress = 0;
+            while ((count = input.read(data)) != -1) {
+                output.write(data, 0, count);
+                progress += count;
+                int percentage = (progress * 100) / fileSize;
+                downloadBar.onProgressUpdated(percentage);
+            }
+        } finally {
+            output.close();
+            input.close();
```
Comment (EmilGeorgiev) [inline src/main/java/downloadAgent/DownloadAgent.java]: use try-with-resources, thereby we will avoid finally block

**Comment by denisdimitrov** [comment]
Now I assert bytes instead of length. And I also used try with resources to avoid the finally block.

**Comment by EmilGeorgiev** [comment]
LGTM

**Pull request merged**